flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -387,10 +393,17 @@ public RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot(
                                         buildSnapshotResult(
                                                 checkpointId,
                                                 delta,
-                                                changelogStateBackendStateCopy)));
+                                                changelogStateBackendStateCopy))
+                        .whenComplete(
+                                (snapshotResult, throwable) ->
+                                        metrics.reportSnapshotResult(snapshotResult))
```

Comment:
This can be executed by a writer thread (upon finishing the upload);
or by the task thread, if the uploads are already finished.

So this call should **probably** be syncrhonized.

The common approach in Flink is to not synchronize metric writes with reads, and allow reporting stale metrics. But here, the update might happen from a different thread, which might lead to inconsistencies (e.g. mixing sizes from different checkpoitns).

Adding synchornization shouldn't incur performance penalty, as this code exeutes infrequently.

So how about marking `reportSnapshotResults` method `synchronized`?